src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs

```diff
@@ -0,0 +1,142 @@
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.Caching.Memory
+{
+    internal partial class CacheEntry
+    {
+        // this type exists just to reduce average CacheEntry size
+        // which typically is not using expiration tokens or callbacks
+        private sealed class CacheEntryTokens
+        {
+            private List<IChangeToken> _expirationTokens;
+            private List<IDisposable> _expirationTokenRegistrations;
+            private List<PostEvictionCallbackRegistration> _postEvictionCallbacks; // this is not really related to tokens, but was moved here to shrink typicall CacheEntry size
+
+            internal List<IChangeToken> ExpirationTokens => _expirationTokens ??= new List<IChangeToken>();
+            internal List<PostEvictionCallbackRegistration> PostEvictionCallbacks => _postEvictionCallbacks ??= new List<PostEvictionCallbackRegistration>();
+
+            internal void AttachTokens()
+            {
+                if (_expirationTokens != null)
+                {
+                    lock (this)
+                    {
+                        for (int i = 0; i < _expirationTokens.Count; i++)
+                        {
+                            IChangeToken expirationToken = _expirationTokens[i];
+                            if (expirationToken.ActiveChangeCallbacks)
+                            {
+                                if (_expirationTokenRegistrations == null)
+                                {
+                                    _expirationTokenRegistrations = new List<IDisposable>(1);
+                                }
+                                IDisposable registration = expirationToken.RegisterChangeCallback(ExpirationCallback, this);
+                                _expirationTokenRegistrations.Add(registration);
+                            }
+                        }
+                    }
+                }
+            }
+
+            internal bool CheckForExpiredTokens(CacheEntry cacheEntry)
+            {
+                if (_expirationTokens != null)
+                {
+                    for (int i = 0; i < _expirationTokens.Count; i++)
+                    {
+                        IChangeToken expiredToken = _expirationTokens[i];
+                        if (expiredToken.HasChanged)
+                        {
+                            cacheEntry.SetExpired(EvictionReason.TokenExpired);
+                            return true;
+                        }
+                    }
+                }
+                return false;
+            }
+
+            internal bool CanCopyTokens() => _expirationTokens != null;
+
+            internal void CopyTokens(CacheEntry parentEntry)
+            {
+                if (_expirationTokens != null)
+                {
+                    lock (this)
+                    {
+                        lock (parentEntry.GetOrCreateTokens())
```

Comment:
Where else is this locked on?

Reply:
https://github.com/dotnet/runtime/pull/45962#discussion_r542216496

Reply:
I'm not sure how that answers my question.  My primary concern would be if it'd ever be possible for something to take the second lock and then try to take the first, creating a priority inversion and deadlock potential.  I'm simply wondering if that's possible.  Does the child/parent nature of this guarantee locks are always acquired in the same order?

Reply:
> Does the child/parent nature of this guarantee locks are always acquired in the same order?

I think it does in the case when the cache item is being added to the cache (by calling the Dispose method):

https://github.com/dotnet/runtime/blob/6e111ea8695635624147227cf685fd6769b88ea4/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L155-L157

but I am not 100% sure if it would not be possible to create a code that does this and abuses the second possible call to this method from the `Cache.TryGetValue` method (which is very unintuitive for me):

https://github.com/dotnet/runtime/blob/6e111ea8695635624147227cf685fd6769b88ea4/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L237-L239

FWIW I have not changed the locking order, it's the same as it used to be so far:

https://github.com/dotnet/runtime/blob/6e111ea8695635624147227cf685fd6769b88ea4/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L331-L340